### ISSUE TITLE
Gracefully handle ALPN unavailability

### ISSUE THREAD

**Issue by washcroft** [mitmproxy/mitmproxy#3210]
mitmproxy is crashing with the below error, not sure how to resolve since OpenSSL has ALPN support and I have recompiled cryptographgy, pyOpenSSL and mitmproxy from source.

# mitmproxy --version
Mitmproxy: 4.0.3
Python:    3.6.3
OpenSSL:   OpenSSL 1.0.2l  25 May 2017
Platform:  Linux-4.9.35-v7+-armv7l-with-debian-8.0

mitmproxy has crashed!

Please lodge a bug report at: https://github.com/mitmproxy/mitmproxy

10.10.10.101:56144: Traceback (most recent call last):
  File "/usr/local/lib/python3.6/site-packages/mitmproxy-4.0.3-py3.6.egg/mitmproxy/proxy/server.py", line 121, in handle
    root_layer()
  File "/usr/local/lib/python3.6/site-packages/mitmproxy-4.0.3-py3.6.egg/mitmproxy/proxy/modes/transparent_proxy.py", line 19, in __call__
    layer()
  File "/usr/local/lib/python3.6/site-packages/mitmproxy-4.0.3-py3.6.egg/mitmproxy/proxy/protocol/tls.py", line 279, in __call__
    self._establish_tls_with_client_and_server()
  File "/usr/local/lib/python3.6/site-packages/mitmproxy-4.0.3-py3.6.egg/mitmproxy/proxy/protocol/tls.py", line 355, in _establish_tls_with_client_and_server
    self._establish_tls_with_server()
  File "/usr/local/lib/python3.6/site-packages/mitmproxy-4.0.3-py3.6.egg/mitmproxy/proxy/protocol/tls.py", line 444, in _establish_tls_with_server
    **args
  File "/usr/local/lib/python3.6/site-packages/mitmproxy-4.0.3-py3.6.egg/mitmproxy/connections.py", line 292, in establish_tls
    self.convert_to_tls(cert=client_cert, sni=sni, **kwargs)
  File "/usr/local/lib/python3.6/site-packages/mitmproxy-4.0.3-py3.6.egg/mitmproxy/net/tcp.py", line 386, in convert_to_tls
    **sslctx_kwargs
  File "/usr/local/lib/python3.6/site-packages/mitmproxy-4.0.3-py3.6.egg/mitmproxy/net/tls.py", line 285, in create_client_context
    **sslctx_kwargs,
  File "/usr/local/lib/python3.6/site-packages/mitmproxy-4.0.3-py3.6.egg/mitmproxy/net/tls.py", line 194, in _create_ssl_context
    context.set_alpn_protos(alpn_protos)
  File "/usr/local/lib/python3.6/site-packages/pyOpenSSL-18.0.0-py3.6.egg/OpenSSL/SSL.py", line 650, in explode
    raise NotImplementedError(error)
NotImplementedError: ALPN not available

**Comment by mhils** [comment]
I'm afraid that's not a mitmproxy bug. What's the output of the following?
```python
from OpenSSL import SSL
print(SSL._lib.Cryptography_HAS_ALPN)
```

**Comment by washcroft** [comment]
@mhils The fact it crashes ungracefully is the bug I was reporting. I've no idea why, but:

> Python 3.6.3 (default, Oct 27 2017, 12:40:04)
> [GCC 4.9.2] on linux
> Type "help", "copyright", "credits" or "license" for more information.
> >>> from OpenSSL import SSL
> >>> print(SSL._lib.Cryptography_HAS_ALPN)
> 0
> >>>

**Comment by mhils** [comment]
Thanks for clarifying. You are very right, we shouldn't crash like this. :)
I unfortunately cannot help you with your actual problem, which is why your pyOpenSSL doesn't ship with ALPN. Can't you just install the manylinux wheel?

**Comment by Kriechi** [comment]
We had many of these checks in the past, but since we moved and depend on newer versions of cryptography/pyopenssl we removed them because by following the dependency graph we SHOULD always have ALPN.

Our dependencies should provide ALPN. If `HAS_ALPN` is 0, then cryptography or pyopenssl are messed up (either local installation or the package itself).

We have to draw the line somewhere - we rely on the "public API" our dependencies provide. We can't double or triple check everything.
